### PR TITLE
feat: new component ColorPicker

### PR DESCRIPTION
**Pull request by RedJue**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |   New Component ColorPicker        |
| 🇨🇳 Chinese |   新组件颜色选择器        |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

### Summary
🎨🧪📝

This pull request adds a new ColorPicker component to the ant-design library, which allows users to select and edit colors in different formats. It includes the component implementation, tests, demos, and documentation files. The component uses the RcColorPicker component as a base, and extends it with some custom logic, styles, and subcomponents. The component also provides a hook to manage the color state, and exports some types for external use.

> _`ColorPicker` shines_
> _A rainbow of props and tests_
> _Spring of components_

### Walkthrough
*  Add the `ColorPicker` component and its documentation ([link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-be92dd92152a780003d48ddc95ec6716367d327f9f890f1221b84614aaa8168fR1-R155), [link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-819859f3d30fc8ec12a8ef1b55ac1d75b4b1d600eb1cee989efd5af57255b560R1-R50), [link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-9d1ff106a032e484c8f982279c4e1295d23f69904a26e020e99310c446e93249R1-R49), [link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-3e9304b2ad2c5648fca4d4743d6d5af269c770f48ddbb29c57c0d901bb516e45R1-R6))
*  Implement the `ExPanel` component to extend the `RcColorPicker` panel with a `ColorDataBar` component ([link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-916c5ee69775c07c3ecb3505f6741803dfbbfc0a93b037af1353e447ec7e41baR1-R21), [link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-9c149a12b5866edb427b9066bd7b0ee0b0091c56142077e14fea6b2f989eaad8R1-R180))
*  Define the `Color` interface and the `ColorFactory` class to wrap the `RcColor` class and provide custom methods and properties ([link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-b05107f431c3fce3a8573df8da8022dbb0a697537ed9ad4b8e674affd5a7c98cR1-R45), [link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-9576ead765b69ad9edf8f148171300a9ac514ef80167a23b0ecd7e0704439647R1-R39))
*  Define the `ColorFormatEnum` type and the `ColorFormat`, `PressetsItem`, and `ColorPickerBaseProps` types for the props and internal components of the `ColorPicker` component ([link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-df67a3a012b615f532d14092a82649a7bf5d7d65773c192377235693e7fd26e0R1-R6), [link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-83dba3026e7dc41ad64d3a2f4d484eeb085f0c3798dc2316fbb90d23520fca6eR1-R16))
*  Implement the `ColorHexInput`, `ColorPlaceholder`, and `ColorSteppers` components to render and edit the color values in different formats ([link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-3ea70913f2f5e38335a70b730457daaa3827d75907311df6df24234f55c31304R1-R45), [link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-836e9e6d03a02a5cdc05b24461a55f692e4dbada93c8b07b3ab4ca5cc5337a07R1-R31), [link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-3fd68486b864ca56f3159c41082c985d14d24297b54639ea2213edd38c7074c0R1-R59))
*  Add the basic usage demo of the `ColorPicker` component and its markdown file ([link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-c9f8c74dc86457938053a55bd0dc68780cb138694afe083e6a5510e26a76f32fR1-R13), [link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-f126c1c22aaeef1ad8e26e6282e140fd953ec34821c205171d80f78d2de8c51cR1-R7))
*  Add the test files for the `ColorPicker` component and its demos, using the testing library and image snapshots ([link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-aa06460351e8630f305cf6509e8137984e0f167ab2560da22874033189234fb4R1-R10), [link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-c9cb9f0026b304e411b8b52dd8bfe190b1e37b96b368edca6856c89c54f9fdeaR1-R5), [link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-de0044e4fa03be11c72b7e52b4a26a86e5563ea6331cf7c47b17b32657db4bbeR1-R3), [link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-028bcdae3bd0c10a9e65ed01c106e4f0045ad92e288354ba40a77734eb651b1dR1-R3))
*  Add the style index file and the style hooks and tokens for the `ColorPicker` component and its subcomponents ([link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-1b0771e050d88aebe327dae413b7f31f0797bcee1515f3c6bb2cd890ba0929d0R1-R275))

### CODE REVIEW

**Comment by MadCcc** [comment]
image
Select 高度，HexInput 字号不对

**Comment by MadCcc** [comment]
image
这里的 handle 被隐藏了

**Comment by MadCcc** [comment]
image
这里应该也要加个阴影

**Comment by fuji221** [comment]
image
1.hex 模式下，输入框内字号不对
2.在popover面板，任意操作几下后，popover会收起

**Comment by fuji221** [comment]
image

- [ ] 默认情况下，描边是灰色，同input
- [ ] hover、click态同input

**Comment by fuji221** [comment]
image
间距为12px

File: components/affix/index.tsx [hunk 1]
```diff
@@ -2,9 +2,9 @@ import classNames from 'classnames';
 import ResizeObserver from 'rc-resize-observer';
 import omit from 'rc-util/lib/omit';
 import React, { createRef, forwardRef, useContext } from 'react';
+import throttleByAnimationFrame from '../_util/throttleByAnimationFrame';
```
Comment (MadCcc) [inline components/affix/index.tsx:5]: 这个文件需要回滚一下

File: components/color-picker/components/ColorPresets.tsx [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    return value;
+  });
+
+const isBright = (value: Color) => {
+  const { r, g, b, a } = value.toRgb();
+  if (a <= 0.5) {
+    return true;
+  }
+  return r * 0.299 + g * 0.587 + b * 0.114 > 192;
+};
```
Comment (MadCcc) [inline components/color-picker/components/ColorPresets.tsx:30]: 直接判断 brightness 就行了吧

File: components/color-picker/components/ColorInput.tsx [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+          >
+            <Select.Option value={ColorFormatEnum.hex}>
+              {ColorFormatEnum.hex.toLocaleUpperCase()}
+            </Select.Option>
+            <Select.Option value={ColorFormatEnum.hsb}>
+              {ColorFormatEnum.hsb.toLocaleUpperCase()}
+            </Select.Option>
+            <Select.Option value={ColorFormatEnum.rgb}>
+              {ColorFormatEnum.rgb.toLocaleUpperCase()}
+            </Select.Option>
```
Comment (MadCcc) [inline components/color-picker/components/ColorInput.tsx:62]: 用 options

File: components/color-picker/enum.ts [hunk 4]
```diff
@@ -0,0 +1,6 @@
+/* eslint-disable import/prefer-default-export */
+export enum ColorFormatEnum {
```
Comment (MadCcc) [inline components/color-picker/enum.ts:2]: 这个写个字面量放 interface 里就行

**Comment by MadCcc** [comment]
- [ ] 鼠标拖动到 popup 外之后松开会直接让 popup 关闭

**Comment by MadCcc** [comment]
image
- [ ] 这个内阴影还没加

**Comment by kiner-tang** [comment]
image
HSB 和 RGB 的默认值是不是展示错了，怎么都展示成了 HEX 了？

File: components/color-picker/ColorPicker.tsx [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+import type { Color } from './color';
+import ColorPlaceholder from './components/ColorPlaceholder';
+import useColorState from './hooks/useColorState';
+import type { ColorFormat, ColorPickerBaseProps, PresetsItem } from './interface';
+import useStyle from './style/index';
+import { customizePrefixCls, generateColor } from './util';
+
+export interface ColorPickerProps
+  extends Omit<RcColorPickerProps, 'onChange' | 'arrow' | 'value' | 'defaultValue' | 'children'> {
+  value?: Color | string;
```
Comment (kiner-tang) [inline components/color-picker/ColorPicker.tsx:22]: 这个类型好多地方都有在用，是不是可以弄一个新类型聚合一下呢？
比如：`type ColorVal = Color | string`，命名上可以再斟酌一下，不过感觉有个别名，在其他地方使用的时候会方便一点

File: components/color-picker/__tests__/index.test.tsx [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import { render } from '@testing-library/react';
+import React from 'react';
+import ColorPicker from '../ColorPicker';
+
+describe('ColorPicker', () => {
```
Comment (kiner-tang) [inline components/color-picker/__tests__/index.test.tsx:5]: 测试用来还没加完吗？

File: components/color-picker/components/ColorPresets.tsx (hunk 2 above)
Comment (RedJue) [reply components/color-picker/components/ColorPresets.tsx:30]: 情况比较多，光明度不足以支撑所有场景，网上找了个通用解法

**Comment by RedJue** [comment]
> image HSB 和 RGB 的默认值是不是展示错了，怎么都展示成了 HEX 了？

demo 之前重构改的忘记修了，已经修复了

File: components/color-picker/__tests__/index.test.tsx (hunk 6 above)
Comment (RedJue) [reply components/color-picker/__tests__/index.test.tsx:5]: 还没，还在UI走查阶段，有空我补一下

File: components/color-picker/ColorPicker.tsx (hunk 5 above)
Comment (RedJue) [reply components/color-picker/ColorPicker.tsx:22]: 可以考虑加个通用类型

File: components/color-picker/demo/format.tsx [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    [colorHex],
+  );
+  const hsbString = useMemo(
+    () => (typeof colorHsb === 'string' ? colorHsb : (colorHsb as Color).toHsbString()),
+    [colorHsb],
+  );
+  const rgbString = useMemo(
+    () => (typeof colorRgb === 'string' ? colorRgb : (colorRgb as Color).toRgbString()),
+    [colorRgb],
+  );
```
Comment (kiner-tang) [inline components/color-picker/demo/format.tsx:21]:
```suggestion
  const hexString = useMemo(
    () => (typeof colorHex === 'string' ? colorHex : colorHex.toHexString()),
    [colorHex],
  );
  const hsbString = useMemo(
    () => (typeof colorHsb === 'string' ? colorHsb : colorHsb.toHsbString()),
    [colorHsb],
  );
  const rgbString = useMemo(
    () => (typeof colorRgb === 'string' ? colorRgb : colorRgb.toRgbString()),
    [colorRgb],
  );
```
这里应该不用类型断言了，前面有 `typeof` 会自动窄化类型的

File: components/color-picker/ColorPicker.tsx [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+      motion={{
+        motionName: getTransitionName(rootPrefixCls, 'slide-up'),
+        motionDeadline: 1000,
+      }}
+      onChange={handleChange}
+      onOpenChange={setPopupOpen}
+      panelRender={(panel) => <ExPanel {...colorBaseProps}>{panel}</ExPanel>}
+      {...extraProps}
+    >
+      {children || (
```
Comment (kiner-tang) [inline components/color-picker/ColorPicker.tsx:154]:
```suggestion
      {children ?? (
```
我在想，这里是不是用 `??` 更合适一点，如果用户就是不想要展示 placeholder ,只想通过 open 控制显示隐藏呢？那如果用户 传一个 false 或 null 过来，依然还是会走后面的默认 placeholder

File: components/color-picker/components/ColorHexInput.tsx [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+  value?: Color;
+  onChange?: (value: Color) => void;
+}
+
+const toHex = (value: string) => value.replace(/[^\w/]/gi, '').slice(0, 8);
+const hexReg = /(^#[\da-f]{6}$)|(^#[\da-f]{8}$)/i;
+
+const ColorHexInput: FC<ColorHexInputProps> = ({ prefixCls, value, onChange }) => {
+  const ColorHexInputPrefixCls = `${prefixCls}-hex-input`;
+  const [hexValue, setHexValue] = useState(toHex(value?.toHexString() || ''));
```
Comment (kiner-tang) [inline components/color-picker/components/ColorHexInput.tsx:18]: 如果 value 有可能为空的话，是不是可以单独定义一个变量处理兜底逻辑，这样就不用后面每次用 value 的时候都用可选链，担心太多可选链会增加包的体积，例如：
```tsx
const val = useMemo(() => value ? value.toHexString() : '', [value])
```
这样后弦直接用 val 就好了？

File: components/color-picker/ColorPicker.tsx (hunk 8 above)
Comment (RedJue) [reply components/color-picker/ColorPicker.tsx:154]: 当初设计上就需要 `trigger` 来触发，最初版本我也是这么想的，后来说了不考虑只显示 `panel`  的场景

File: components/color-picker/components/ColorHexInput.tsx (hunk 9 above)
Comment (RedJue) [reply components/color-picker/components/ColorHexInput.tsx:18]: 可以优化一下，可选链多了确实不好

File: components/color-picker/demo/format.tsx (hunk 7 above)
Comment (RedJue) [reply components/color-picker/demo/format.tsx:21]: 确实，写多了，我优化一下

File: components/color-picker/ColorPicker.tsx (hunk 8 above)
Comment (kiner-tang) [reply components/color-picker/ColorPicker.tsx:154]: ok

File: components/color-picker/components/ColorHexInput.tsx [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+  onChange?: (value: Color) => void;
+}
+
+const toHex = (value: string) => value.replace(/[^\w/]/gi, '').slice(0, 8);
+const hexReg = /(^#[\da-f]{6}$)|(^#[\da-f]{8}$)/i;
+const getHexString = (value?: Color) => (value ? toHex(value.toHexString()) : '');
+
+const ColorHexInput: FC<ColorHexInputProps> = ({ prefixCls, value, onChange }) => {
+  const ColorHexInputPrefixCls = `${prefixCls}-hex-input`;
+  const [hexValue, setHexValue] = useState(getHexString(value));
```
Comment (MadCcc) [inline components/color-picker/components/ColorHexInput.tsx:19]: 不如让 Color 支持 toHex

Comment (RedJue) [reply components/color-picker/components/ColorHexInput.tsx:19]: 可以加一个，这个**API**需要暴露出去么

**Comment by li-jia-nan** [comment]
Conflicting

File: components/color-picker/ExPanel.tsx [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+import ColorClear from './components/ColorClear';
+import ColorInput from './components/ColorInput';
+import ColorPresets from './components/ColorPresets';
+import type { ColorPickerBaseProps } from './interface';
+
+interface ExPanelProps extends ColorPickerBaseProps {
+  children?: React.ReactElement;
+}
+
+const ExPanel: FC<ExPanelProps> = (props) => {
```
Comment (MadCcc) [inline components/color-picker/ExPanel.tsx:13]:
```suggestion
const ColorPickerPanel: FC<ExPanelProps> = (props) => {
```
命名要明确一点

File: components/color-picker/components/ColorSteppers.tsx [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+
+  // Update step value
+  useEffect(() => {
+    if (isNumber(value)) {
+      setStepValue(value);
+    }
+  }, [value]);
+
+  return (
+    <div className={classNames(ColorSteppersPrefixCls, className)}>
```
Comment (MadCcc) [inline components/color-picker/components/ColorSteppers.tsx:40]: 这层是必要的么？

File: components/color-picker/components/ColorSteppers.tsx [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+  max?: number;
+  onChange?: (value: number | null) => void;
+  className?: string;
+  prefix?: (prefixCls: string) => React.ReactNode;
+  formatter?: InputNumberProps<number>['formatter'];
+}
+
+const isNumber = (value: number | undefined) => typeof value === 'number' && !isNaN(value);
+
+const ColorSteppers: FC<ColorSteppersProps> = ({
```
Comment (MadCcc) [inline components/color-picker/components/ColorSteppers.tsx:20]: 这整个组件封装的成分不是很多，大部分属性都是透传，感觉不是很有必要，只用 InputNumber + className 就可以的话可以简化掉了

Comment (RedJue) [reply components/color-picker/components/ColorSteppers.tsx:20]: 具体看下逻辑，应该也可以

**Comment by MadCcc** [comment]
Kapture 2023-05-06 at 10 04 30
反向之后动画方向没有反向。
这个省力一点的话就可以让 rc-color-picker 把 popup 内的面板 export 出来，然后在 antd 层用 Popover 承载，就不用关心这些弹层相关的逻辑了。

**Comment by RedJue** [comment]
> Kapture 2023-05-06 at 10 04 30
> 
> 反向之后动画方向没有反向。
> 
> 这个省力一点的话就可以让 rc-color-picker 把 popup 内的面板 export 出来，然后在 antd 层用 Popover 承载，就不用关心这些弹层相关的逻辑了。

嗯 确实 我也想过这个 感觉重新实现一遍popover 没啥必要 这个我优化一下吧 antd 里面能复用尽量复用

**Comment by MadCcc** [comment]
image
这个 demo 建议先按设计稿上的来，标题应该是 “自定义触发器”

**Comment by MadCcc** [comment]
文档里也列一下 Color 支持的属性

**Comment by RedJue** [comment]
> image
> 
> 这个 demo 建议先按设计稿上的来，标题应该是 “自定义触发器”

好 之前说不要出现自定义关键词就把这些去掉了😢

**Comment by MadCcc** [comment]
> > image
> > 这个 demo 建议先按设计稿上的来，标题应该是 “自定义触发器”
> 
> 好 之前说不要出现自定义关键词就把这些去掉了😢

正常的属性不需要“自定义“的字眼，但是这里就确实是自定义

**Comment by RedJue** [comment]
> > > image
> 
> > > 这个 demo 建议先按设计稿上的来，标题应该是 “自定义触发器”
> 
> > 
> 
> > 好 之前说不要出现自定义关键词就把这些去掉了😢
> 
> 
> 
> 正常的属性不需要“自定义“的字眼，但是这里就确实是自定义

了解了👌

**Comment by afc163** [comment]
> 嗯 确实 我也想过这个 感觉重新实现一遍popover 没啥必要 这个我优化一下吧 antd 里面能复用尽量复用

+1 ，外框整个就是 Popover，不需要重新实现。

**Comment by RedJue** [comment]
> > 嗯 确实 我也想过这个 感觉重新实现一遍popover 没啥必要 这个我优化一下吧 antd 里面能复用尽量复用
> 
> 
> 
> +1 ，外框整个就是 Popover，不需要重新实现。

嗯嗯 我重新改一下

File: components/color-picker/ColorPicker.tsx [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+    format,
+    direction,
+    onFormatChange,
+    updateColor,
+    updateClearColor,
+  };
+
+  return wrapSSR(
+    <Popover
+      overlayClassName={rootClassName}
```
Comment (MadCcc) [inline components/color-picker/ColorPicker.tsx:125]: 这个应该给 Popover 的 `rootClassName` 吧

File: components/color-picker/ColorPicker.tsx [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+    color: colorValue,
+    allowClear,
+    clearColor,
+    disabled,
+    presets,
+    format,
+    direction,
+    onFormatChange,
+    updateColor,
+    updateClearColor,
```
Comment (MadCcc) [inline components/color-picker/ColorPicker.tsx:120]: `onChange` 和 `onClear` 更好些

File: components/color-picker/ColorPicker.tsx [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+import useColorState from './hooks/useColorState';
+import type { ColorFormat, ColorPickerBaseProps, PresetsItem } from './interface';
+import useStyle from './style/index';
+import { customizePrefixCls, generateColor } from './util';
+
+export interface ColorPickerProps
+  extends Omit<
+    RcColorPickerProps,
+    'onChange' | 'arrow' | 'value' | 'defaultValue' | 'children' | 'panelRender'
+  > {
```
Comment (MadCcc) [inline components/color-picker/ColorPicker.tsx:23]: 这里类型不该继承 RcColorPickerProps 了，应该继承 RcColorPickerPanelProps

File: components/color-picker/ColorPicker.tsx [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+    <Popover
+      overlayClassName={rootClassName}
+      style={styles?.popup}
+      onOpenChange={setPopupOpen}
+      content={
+        <RcColorPickerPanel
+          prefixCls={prefixCls}
+          value={colorValue.toHsb()}
+          onChange={handleChange}
+          panelRender={(panel) => <ColorPickerPanel {...colorBaseProps}>{panel}</ColorPickerPanel>}
```
Comment (MadCcc) [inline components/color-picker/ColorPicker.tsx:133]: 只用 Panel 就不太需要 panelRender，直接在 ColorPickerPanel 里引用 RcColorPickerPanel 就可以

File: components/color-picker/ColorPickerPanel.tsx [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+    prefixCls,
+    children,
+    allowClear,
+    presets,
+    updateColor,
+    updateClearColor,
+    color,
+    ...injectProps
+  } = props;
+  const ColorPickerPanelPrefixCls = `${prefixCls}-inner-panel`;
```
Comment (MadCcc) [inline components/color-picker/ColorPickerPanel.tsx:24]:
```suggestion
  const ColorPickerPanelPrefixCls = `${prefixCls}-panel`;
```
短一点

File: components/color-picker/ColorPicker.tsx (hunk 15 above)
Comment (MadCcc) [reply components/color-picker/ColorPicker.tsx:120]: 命名规范参考：https://github.com/ant-design/ant-design/wiki/API-Naming-rules

File: components/color-picker/components/ColorPresets.tsx [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+  value: color,
+  onChange,
+  direction,
+}) => {
+  const [locale] = useLocale('ColorPicker');
+  const [presetsValue] = useMergedState(genPresetColor(presets), {
+    value: genPresetColor(presets),
+    onChange: (item) => genPresetColor(item),
+  });
+  const ColorPresetsPrefixCls = `${prefixCls}-presets`;
```
Comment (MadCcc) [inline components/color-picker/components/ColorPresets.tsx:45]:
```suggestion
  const colorPresetsPrefixCls = `${prefixCls}-presets`;
```
变量小驼峰

File: components/color-picker/components/ColorPresets.tsx [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+                    key={`preset-${presetColor.toHexString()}`}
+                    className={classNames(`${ColorPresetsPrefixCls}-color`, {
+                      [`${ColorPresetsPrefixCls}-color-checked`]:
+                        presetColor.toHexString() === color?.toHexString(),
+                    })}
+                    onClick={() => handleClick(presetColor)}
+                  >
+                    <div
+                      className={classNames(`${ColorPresetsPrefixCls}-color-box`, {
+                        [`${ColorPresetsPrefixCls}-color-box-rtl`]: direction === 'rtl',
```
Comment (MadCcc) [inline components/color-picker/components/ColorPresets.tsx:77]: rtl 尽量用 logicalProperties 解决，非必要不添加 class。

File: components/color-picker/components/ColorPresets.tsx [hunk 21]
```diff
@@ -0,0 +1,102 @@
...
+                        [`${ColorPresetsPrefixCls}-color-bright`]: isBright(presetColor),
+                        [`${ColorPresetsPrefixCls}-color-dark`]: !isBright(presetColor),
```
Comment (MadCcc) [inline components/color-picker/components/ColorPresets.tsx:79]: 两个留一个即可

File: components/color-picker/components/ColorPresets.tsx [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+                      })}
+                    >
+                      <div className={`${ColorPresetsPrefixCls}-color-block`} />
```
Comment (MadCcc) [inline components/color-picker/components/ColorPresets.tsx:82]: 这个空节点是干什么的？

File: components/color-picker/interface.ts [hunk 23]
```diff
@@ -0,0 +1,28 @@
...
+  clearColor?: boolean;
+  disabled?: boolean;
+  direction?: DirectionType;
+  presets?: PresetsItem[];
+  onFormatChange?: ColorPickerProps['onFormatChange'];
+  updateColor?: (value?: Color) => void;
+  updateClearColor?: (clear?: boolean) => void;
+}
+
+export enum ColorFormatEnum {
```
Comment (MadCcc) [inline components/color-picker/interface.ts:24]: 这个 enum 冗余了，没有必要。代码里用 ColorFormat 就够了

File: components/color-picker/ColorPickerPanel.tsx (hunk 18 above)
Comment (RedJue) [reply components/color-picker/ColorPickerPanel.tsx:24]: 和 rc 里面 panel 命名冲突了，所以加了个 inner 区分一下。

File: components/color-picker/components/ColorPresets.tsx (hunk 20 above)
Comment (RedJue) [reply components/color-picker/components/ColorPresets.tsx:77]: 由于图标旋转的问题，必须得加 ` direction: 'ltr',`，所以加了个类名区别

File: components/color-picker/ColorPicker.tsx (hunk 16 above)
Comment (RedJue) [reply components/color-picker/ColorPicker.tsx:23]: `RcColorPickerPanelProps` 缺失很多属性定义，这些都是定义在 `RcColorPickerProps ` 上

File: components/color-picker/components/ColorPresets.tsx (hunk 22 above)
Comment (RedJue) [reply components/color-picker/components/ColorPresets.tsx:82]: 样式节点，具体展现为 `size`为 18的小方块。

File: components/color-picker/components/ColorPresets.tsx (hunk 20 above)
Comment (MadCcc) [reply components/color-picker/components/ColorPresets.tsx]: rlt 加在最外面

File: components/color-picker/ColorPicker.tsx (hunk 16 above)
Comment (MadCcc) [reply components/color-picker/ColorPicker.tsx:23]: 缺失的类型定义在 antd 里补充，原则上用什么组件就 extend 这个组件的 props。RcColorPickerProps 的属性大多都是没有消费到的。

File: components/color-picker/components/ColorSteppers.tsx [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+}) => {
+  const colorSteppersPrefixCls = `${prefixCls}-steppers`;
+  const [stepValue, setStepValue] = useState(value);
+
+  // Update step value
+  useEffect(() => {
+    if (isNumber(value)) {
+      setStepValue(value);
+    }
+  }, [value]);
```
Comment (MadCcc) [inline components/color-picker/components/ColorSteppers.tsx:37]: 感觉没啥必要这样处理，这应该都是 InputNumber 的内部逻辑

File: components/color-picker/index.en-US.md [hunk 25]
```diff
@@ -0,0 +1,67 @@
+---
+category: Components
+title: ColorPicker
+cover: https://mdn.alipayobjects.com/huamei_7uahnr/afts/img/A*cJopQrf0ncwAAAAAAAAAAAAADrJ8AQ/original
+coverDark: https://mdn.alipayobjects.com/huamei_7uahnr/afts/img/A*M4PBTZ_n9OgAAAAAAAAAAAAADrJ8AQ/original
+demo:
+  cols: 2
+group:
+  title: Data Entry
+  order: 5
```
Comment (MadCcc) [inline components/color-picker/index.en-US.md:10]:
```suggestion
  order: 5
```

**Comment by MadCcc** [comment]
image
能不透传还是不透传，除非是套了一层的组件。HTML 元素上使用 spread 也要注意下。

**Comment by li-jia-nan** [comment]
闲夕大佬这么晚还在推代码

**Comment by RedJue** [comment]
辛苦了大佬🙏

**Comment by li-jia-nan** [comment]
看样子可以上 5.5 的车了🎉

**Comment by fuji221** [comment]
image
popover在上方时间距有问题

**Comment by MadCcc** [comment]
argos 没有截图，还得看看

**Comment by fuji221** [comment]
> 

没问题了

**Review by MadCcc — APPROVED** [review]

**Comment by 1123612483** [comment]
能否给组件加上吸管拾色功能？
https://developer.mozilla.org/en-US/docs/Web/API/EyeDropper_API
image

**Comment by afc163** [comment]
图片

弹出动画方向有点不对，上面这个是从中间弹开的，应该从箭头方向弹开。

**Comment by afc163** [comment]
图片

操作点没有出息鼠标手型。

**Comment by afc163** [comment]
图片

切换无效。

**Comment by afc163** [comment]
图片

少 。

**Comment by afc163** [comment]
```
import type { Color } from 'antd/es/color-picker';
```

用 es 目录。

**Comment by afc163** [comment]
图片

这个图案是重复的，感觉 16X16 + repeat 就够了，不需要这么大。

**Pull request merged**